rate_limit_quotas.py

```diff
@@ -0,0 +1,186 @@
+    * late limiting based on resource usage (time spent on API calls)
+    * bucketing based on day, week, month
+    * statistics
+    * metrics
+    * fast in-memory storage
+
+Also provides a connector for Flask
+
+"""
+
+import time
+import ipaddress
+from typing import Dict
+
+
+class Limiter:
+    def __init__(
+        self,
+        limits: dict,
+        token_check_callback=None,
+        ipaddr_methods=["X-Real-Ip", "socket"],
+    ):
+        # Bucket sequence: month, week, day
+        self._ipaddr_limits = [
+            limits.get(l, None)
+            for l in ("ipaddr_per_month", "ipaddr_per_week", "ipaddr_per_day")
+        ]
+        self._token_limits = [
+            limits.get(l, None)
+            for l in ("token_per_month", "token_per_week", "token_per_day")
+        ]
+        self._ipaddr_buckets = ({}, {}, {})  # type: Dict
+        self._token_buckets = ({}, {}, {})  # type: Dict
+        self._token_check_callback = token_check_callback
+        self._ipaddr_extraction_methods = ipaddr_methods
+        self._last_quota_update_time = time.monotonic()
+        self.increment_quota_counters(1)
+        self.refresh_quota_counters_if_needed()
+
+    def increment_quota_counters(self, tdelta: int):
+        """Delta: time from previous run in seconds"""
+        if tdelta <= 0:
+            return
+
+        iterable = (
+            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (7 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (1 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (30 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (7 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+            (1 * 24, self._ipaddr_limits[0], self._ipaddr_buckets[0]),
+        )
+        for hours, limit, bucket in iterable:
+            vdelta = limit / hours / 3600 * tdelta
+            to_delete = []
+            for k, v in bucket.items():
+                v += vdelta
+                if v >= limit:
+                    to_delete.append(k)
+                else:
+                    bucket[k] = v
+
+            for k in to_delete:
+                del bucket[k]
+
+    def refresh_quota_counters_if_needed(self):
+        t = time.monotonic()
+        delta = t - self._last_quota_update_time
+        if delta > 3600:
+            self.increment_quota_counters(delta)
+
+        self._last_quota_update_time = t
+
+    def consume_quota(self, elapsed: float, ipaddr=None, token=None) -> None:
+        """Consume quota in seconds
+        """
+        if ipaddr:
+            for n, limit in enumerate(self._ipaddr_limits):
+                b = self._ipaddr_buckets[n]
+                b[ipaddr] = b.get(ipaddr, limit) - elapsed
+
+        print([b[ipaddr] for b in self._ipaddr_buckets])
+        self.increment_quota_counters(1)
+
+    def get_minimum_across_quotas(self, ipaddr=None, token=None) -> int:
+        if ipaddr:
```

Comment:
Should we maybe handle the case in which `ipaddr` is not set? Is there any way this function could get called without a valid `ipaddr` value?

Reply:
Not a risk with the current codebase, but I'm adding assertions around this.